curvefs/src/mds/schedule/topoAdapter.cpp

```diff
@@ -0,0 +1,349 @@
+
+bool MetaServerInfo::IsUnstable() const {
+    return state == OnlineState::UNSTABLE;
+}
+
+bool MetaServerInfo::IsHealthy() const { return state == OnlineState::ONLINE; }
+
+TopoAdapterImpl::TopoAdapterImpl(std::shared_ptr<Topology> topo,
+                                 std::shared_ptr<TopologyManager> manager) {
+    this->topo_ = topo;
+    this->topoManager_ = manager;
+}
+
+std::vector<PoolIdType> TopoAdapterImpl::Getpools() {
+    return topo_->GetPoolInCluster();
+}
+
+bool TopoAdapterImpl::GetCopySetInfo(const CopySetKey &id, CopySetInfo *info) {
+    ::curvefs::mds::topology::CopySetInfo csInfo;
+    // cannot get copyset info
+    if (!topo_->GetCopySet(id, &csInfo)) {
+        return false;
+    }
+
+    // cannot get logical pool
+    ::curvefs::mds::topology::Pool lpool;
+    if (!topo_->GetPool(csInfo.GetPoolId(), &lpool)) {
+        return false;
+    }
+
+    if (!CopySetFromTopoToSchedule(csInfo, info)) {
+        return false;
+    }
+
+    info->poolWork = lpool.GetPoolAvaliableFlag();
+    return true;
+}
+
+std::vector<CopySetInfo> TopoAdapterImpl::GetCopySetInfos() {
+    std::vector<CopySetInfo> infos;
+    for (auto copySetKey : topo_->GetCopySetsInCluster()) {
+        CopySetInfo copySetInfo;
+        if (GetCopySetInfo(copySetKey, &copySetInfo)) {
+            if (copySetInfo.poolWork) {
+                infos.push_back(copySetInfo);
+            }
+        }
+    }
+    return infos;
+}
+
+std::vector<CopySetInfo> TopoAdapterImpl::GetCopySetInfosInMetaServer(
+    MetaServerIdType id) {
+    std::vector<CopySetKey> keys = topo_->GetCopySetsInMetaServer(id);
+
+    std::vector<CopySetInfo> out;
+    for (auto key : keys) {
+        CopySetInfo info;
+        if (GetCopySetInfo(key, &info)) {
+            if (info.poolWork) {
+                out.emplace_back(info);
+            }
+        }
+    }
+    return out;
+}
+
+bool TopoAdapterImpl::GetMetaServerInfo(MetaServerIdType id,
+                                        MetaServerInfo *out) {
+    assert(out != nullptr);
+
+    ::curvefs::mds::topology::MetaServer ms;
+    if (!topo_->GetMetaServer(id, &ms)) {
+        LOG(ERROR) << "can not get metaServer:" << id << " from topology";
+        return false;
+    }
+    return MetaServerFromTopoToSchedule(ms, out);
+}
+
+std::vector<MetaServerInfo> TopoAdapterImpl::GetMetaServerInfos() {
+    std::vector<MetaServerInfo> infos;
+    for (auto metaServerId : topo_->GetMetaServerInCluster(
+             [](const MetaServer &ms) { return true; })) {
+        MetaServerInfo info;
+        if (GetMetaServerInfo(metaServerId, &info)) {
+            infos.push_back(info);
+        }
+    }
+
+    return infos;
+}
+
+std::vector<MetaServerInfo> TopoAdapterImpl::GetMetaServersInPool(
+    PoolIdType poolId) {
+    std::vector<MetaServerInfo> infos;
+    auto ids = topo_->GetMetaServerInPool(
+        poolId, [](const MetaServer &metaserver) { return true; });
```

Comment:
the second parameter is unnecessary, function default parameter already return true by default

Reply:
done